crates/optimism/txpool/src/validator.rs

```diff
@@ -225,14 +209,81 @@ where
         &self,
         transactions: Vec<(TransactionOrigin, Tx)>,
     ) -> Vec<TransactionValidationOutcome<Tx>> {
-        futures_util::future::join_all(
-            transactions.into_iter().map(|(origin, tx)| self.validate_one(origin, tx)),
-        )
+        let transactions = transactions
+            .into_iter()
+            .map(|(origin, tx)| {
+                self.apply_op_checks_no_state(tx)
+                    .and_then(|tx| self.inner.validate_one_no_state(origin, tx))
+                    .map(|tx| (origin, tx))
+            })
+            .collect::<Vec<_>>();
+
+        // bail early if all transactions failed validation without state
+        let some_pass = transactions.iter().any(|res| res.is_ok());
+        if !some_pass {
+            return transactions.into_iter().filter_map(|res| res.err()).collect()
+        }
+
+        // otherwise load state from DB for checks against state
+        let state: Arc<dyn StateProvider> = match self.client().latest() {
+            Ok(s) => s.into(),
+            Err(err) => {
+                return transactions
+                    .into_iter()
+                    .map(|res| match res {
+                        Ok((_, tx)) => {
+                            TransactionValidationOutcome::Error(*tx.hash(), Box::new(err.clone()))
+                        }
+                        Err(already_invalid) => already_invalid,
+                    })
+                    .collect()
+            }
+        };
+
+        futures_util::future::join_all(transactions.into_iter().map(async |res| match res {
+            Ok((origin, tx)) => {
+                let outcome = self.inner.validate_one_against_state(origin, tx, state.clone());
+                self.apply_op_checks_against_state(outcome).await
+            }
+            Err(invalid_outcome) => invalid_outcome,
+        }))
         .await
     }
 
-    /// Performs the necessary opstack specific checks based on top of the regular eth outcome.
-    fn apply_op_checks(
+    /// Applies OP validity checks, that do _not_ require reading latest state from DB (or from
+    /// Superchain oracle), to single transaction.
+    ///
+    /// Returns the unmodified transaction if all checks pass.
+    ///
+    /// Applies OP-protocol specific checks:
+    /// - ensures tx is not eip4844
+    pub fn apply_op_checks_no_state(
+        &self,
+        transaction: Tx,
+    ) -> Result<Tx, TransactionValidationOutcome<Tx>> {
```

Comment:
possibly makes sense for this to return `ValidTransaction<Tx>` instead of just `Tx`